File: PPMLib/Extensions/BinaryReaderExtensions.cs
using System.IO;
using System.Text;

namespace PPMLib.Extensions
{
    internal static class BinaryReaderExtensions
    {
        public static string ReadWChars(this BinaryReader br, int count)
            => Encoding.Unicode.GetString(br.ReadBytes(2 * count));

        public static PPMFilename ReadPPMFilename(this BinaryReader br)
            => new PPMFilename(br.ReadBytes(18));

        public static PPMFileFragment ReadPPMFileFragment(this BinaryReader br)
            => new PPMFileFragment(br.ReadBytes(8));

        public static PPMTimestamp ReadPPMTimestamp(this BinaryReader br)
            => new PPMTimestamp(br.ReadUInt32());

        public static PPMThumbnail ReadPPMThumbnail(this BinaryReader br)
            => new PPMThumbnail(br.ReadBytes(1536));

        public static PPMFrame ReadPPMFrame(this BinaryReader br)
        {
            PPMFrame frame = new PPMFrame();
            frame._firstByteHeader = br.ReadByte();
            if ((frame._firstByteHeader & 0x60) != 0)
            {
                frame._translateX = br.ReadSByte();
                frame._translateY = br.ReadSByte();
            }

            frame.PaperColor = (PaperColor)(frame._firstByteHeader % 2);
            frame.Layer1.PenColor = (PenColor)((frame._firstByteHeader >> 1) & 3);
            frame.Layer2.PenColor = (PenColor)((frame._firstByteHeader >> 3) & 3);

            frame.Layer1._linesEncoding = br.ReadBytes(0x30);
            frame.Layer2._linesEncoding = br.ReadBytes(0x30);

            for (int y = 0, yy; y < 192; y++)
            {
                yy = y << 5;
                switch (frame.Layer1.LinesEncoding(y))
                {
                    case 0:
                        break;
                    case (LineEncoding)1:
                        for (int x = 0; x < 32; x++) frame.Layer1._layerData[yy + x] = 0x00;
                        byte b1 = br.ReadByte(), b2 = br.ReadByte(), b3 = br.ReadByte(), b4 = br.ReadByte();
                        uint bytes = ((uint)(b1 << 24)) + ((uint)(b2 << 16)) + ((uint)(b3 << 8)) + b4;
                        while (bytes != 0)
                        {
                            if ((bytes & 0x80000000) != 0)
                                frame.Layer1._layerData[yy] = br.ReadByte();
                            bytes <<= 1;
                            yy++;
                        }
                        break;
                    case (LineEncoding)2:
                        for (int x = 0; x < 32; x++) frame.Layer1._layerData[yy + x] = 0xFF;
                        b1 = br.ReadByte(); b2 = br.ReadByte(); b3 = br.ReadByte(); b4 = br.ReadByte();
                        bytes = ((uint)(b1 << 24)) + ((uint)(b2 << 16)) + ((uint)(b3 << 8)) + b4;
                        while (bytes != 0)
                        {
                            if ((bytes & 0x80000000) != 0)
                                frame.Layer1._layerData[yy] = br.ReadByte();
                            bytes <<= 1;
                            yy++;
                        }
                        break;
                    case (LineEncoding)3:
                        for (int x = 0; x < 32; x++)
                            frame.Layer1._layerData[yy + x] = br.ReadByte();
                        break;
                }
            }
            for (int y = 0, yy; y < 192; y++)
            {
                yy = y << 5;
                switch (frame.Layer2.LinesEncoding(y))
                {
                    case 0:
                        break;
                    case (LineEncoding)1:
                        for (int x = 0; x < 32; x++) frame.Layer2._layerData[yy + x] = 0x00;
                        byte b1 = br.ReadByte(), b2 = br.ReadByte(), b3 = br.ReadByte(), b4 = br.ReadByte();
                        uint bytes = ((uint)(b1 << 24)) + ((uint)(b2 << 16)) + ((uint)(b3 << 8)) + b4;
                        while (bytes != 0)
                        {
                            if ((bytes & 0x80000000) != 0)
                                frame.Layer2._layerData[yy] = br.ReadByte();
                            bytes <<= 1;
                            yy++;
                        }
                        break;
                    case (LineEncoding)2:
                        for (int x = 0; x < 32; x++) frame.Layer2._layerData[yy + x] = 0xFF;
                        b1 = br.ReadByte(); b2 = br.ReadByte(); b3 = br.ReadByte(); b4 = br.ReadByte();
                        bytes = ((uint)(b1 << 24)) + ((uint)(b2 << 16)) + ((uint)(b3 << 8)) + b4;
                        while (bytes != 0)
                        {
                            if ((bytes & 0x80000000) != 0)
                                frame.Layer2._layerData[yy] = br.ReadByte();
                            bytes <<= 1;
                            yy++;
                        }
                        break;
                    case (LineEncoding)3:
                        for (int x = 0; x < 32; x++)
                            frame.Layer2._layerData[yy + x] = br.ReadByte();
                        break;
                }
            }
            return frame;
        }
    }
}

File: Extensions/BinaryReaderExtensions.vb
Imports System.IO
Imports System.Text

Namespace PPMLib.Extensions
	Friend Module BinaryReaderExtensions
		<System.Runtime.CompilerServices.Extension> _
		Public Function ReadWChars(ByVal br As BinaryReader, ByVal count As Integer) As String
			Return Encoding.Unicode.GetString(br.ReadBytes(2 * count))
		End Function

		<System.Runtime.CompilerServices.Extension> _
		Public Function ReadPPMFilename(ByVal br As BinaryReader) As PPMFilename
			Return New PPMFilename(br.ReadBytes(18))
		End Function

		<System.Runtime.CompilerServices.Extension> _
		Public Function ReadPPMFileFragment(ByVal br As BinaryReader) As PPMFileFragment
			Return New PPMFileFragment(br.ReadBytes(8))
		End Function

		<System.Runtime.CompilerServices.Extension> _
		Public Function ReadPPMTimestamp(ByVal br As BinaryReader) As PPMTimestamp
			Return New PPMTimestamp(br.ReadUInt32())
		End Function

		<System.Runtime.CompilerServices.Extension> _
		Public Function ReadPPMThumbnail(ByVal br As BinaryReader) As PPMThumbnail
			Return New PPMThumbnail(br.ReadBytes(1536))
		End Function

		<System.Runtime.CompilerServices.Extension> _
		Public Function ReadPPMFrame(ByVal br As BinaryReader) As PPMFrame
			Dim frame As New PPMFrame()
			frame._firstByteHeader = br.ReadByte()
			If (frame._firstByteHeader And &H60) <> 0 Then
				frame._translateX = br.ReadSByte()
				frame._translateY = br.ReadSByte()
			End If

			frame.PaperColor = CType(frame._firstByteHeader Mod 2, PaperColor)
			frame.Layer1.PenColor = CType((frame._firstByteHeader >> 1) And 3, PenColor)
			frame.Layer2.PenColor = CType((frame._firstByteHeader >> 3) And 3, PenColor)

			frame.Layer1._linesEncoding = br.ReadBytes(&H30)
			frame.Layer2._linesEncoding = br.ReadBytes(&H30)

			Dim y As Integer = 0
			Dim yy As Integer
			Do While y < 192
				yy = y << 5
				Select Case frame.Layer1.LinesEncoding(y)
					Case 0
					Case CType(1, LineEncoding)
						For x As Integer = 0 To 31
							frame.Layer1._layerData(yy + x) = &H0
						Next x
						Dim b1 As Byte = br.ReadByte(), b2 As Byte = br.ReadByte(), b3 As Byte = br.ReadByte(), b4 As Byte = br.ReadByte()
						Dim bytes As UInteger = (CUInt(b1 << 24)) + (CUInt(b2 << 16)) + (CUInt(b3 << 8)) + b4
						Do While bytes <> 0
							If (bytes And &H80000000UI) <> 0 Then
								frame.Layer1._layerData(yy) = br.ReadByte()
							End If
							bytes <<= 1
							yy += 1
						Loop
					Case CType(2, LineEncoding)
						For x As Integer = 0 To 31
							frame.Layer1._layerData(yy + x) = &HFF
						Next x
						b1 = br.ReadByte()
						b2 = br.ReadByte()
						b3 = br.ReadByte()
						b4 = br.ReadByte()
						bytes = (CUInt(b1 << 24)) + (CUInt(b2 << 16)) + (CUInt(b3 << 8)) + b4
						Do While bytes <> 0
							If (bytes And &H80000000UI) <> 0 Then
								frame.Layer1._layerData(yy) = br.ReadByte()
							End If
							bytes <<= 1
							yy += 1
						Loop
					Case CType(3, LineEncoding)
						For x As Integer = 0 To 31
							frame.Layer1._layerData(yy + x) = br.ReadByte()
						Next x
				End Select
				y += 1
			Loop
			y = 0
			Dim yy As Integer
			Do While y < 192
				yy = y << 5
				Select Case frame.Layer2.LinesEncoding(y)
					Case 0
					Case CType(1, LineEncoding)
						For x As Integer = 0 To 31
							frame.Layer2._layerData(yy + x) = &H0
						Next x
						Dim b1 As Byte = br.ReadByte(), b2 As Byte = br.ReadByte(), b3 As Byte = br.ReadByte(), b4 As Byte = br.ReadByte()
						Dim bytes As UInteger = (CUInt(b1 << 24)) + (CUInt(b2 << 16)) + (CUInt(b3 << 8)) + b4
						Do While bytes <> 0
							If (bytes And &H80000000UI) <> 0 Then
								frame.Layer2._layerData(yy) = br.ReadByte()
							End If
							bytes <<= 1
							yy += 1
						Loop
					Case CType(2, LineEncoding)
						For x As Integer = 0 To 31
							frame.Layer2._layerData(yy + x) = &HFF
						Next x
						b1 = br.ReadByte()
						b2 = br.ReadByte()
						b3 = br.ReadByte()
						b4 = br.ReadByte()
						bytes = (CUInt(b1 << 24)) + (CUInt(b2 << 16)) + (CUInt(b3 << 8)) + b4
						Do While bytes <> 0
							If (bytes And &H80000000UI) <> 0 Then
								frame.Layer2._layerData(yy) = br.ReadByte()
							End If
							bytes <<= 1
							yy += 1
						Loop
					Case CType(3, LineEncoding)
						For x As Integer = 0 To 31
							frame.Layer2._layerData(yy + x) = br.ReadByte()
						Next x
				End Select
				y += 1
			Loop
			Return frame
		End Function
	End Module
End Namespace
